h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,168 @@
+    // to use CategoricalWrappedVec.computeMap() since indexes for levels could differ in both frames for the same level values.
+    private String _levelValue;
+    private int _foldValue;
+
+    CompositeLookupKey(String levelValue, int fold) {
+      this._levelValue = levelValue;
+      this._foldValue = fold;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      CompositeLookupKey lookupKey = (CompositeLookupKey) o;
+      return _foldValue == lookupKey._foldValue &&
+              _levelValue.equals(lookupKey._levelValue);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(_levelValue, _foldValue);
+    }
+  }
+
+  static class EncodingData {
+    private long _numerator;
+    private long _denominator;
+    
+    EncodingData(long numerator, long denominator) {
+      this._numerator = numerator;
+      this._denominator = denominator;
+    }
+
+    long getNumerator() {
+      return _numerator;
+    }
+
+    long getDenominator() {
+      return _denominator;
+    }
+  }
+
+  private static String FOLD_VALUE_IS_OUT_OF_RANGE_MSG = "Fold value should be a non-negative integer (i.e. should belong to [0, Integer.MAX_VALUE] range)";
+
+
+  static class FrameWithEncodingDataToHashMap extends MRTask<FrameWithEncodingDataToHashMap> {
+
+    IcedHashMap<CompositeLookupKey, EncodingData> _encodingDataMapPerNode = new IcedHashMap<>();
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx;
+
+    IcedHashMap<CompositeLookupKey, EncodingData> _encodingDataMapFinal = new IcedHashMap<>();
+
+    FrameWithEncodingDataToHashMap(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx) {
+      this._categoricalColumnIdx = categoricalColumnIdx;
+      this._foldColumnIdx = foldColumnId;
+      this._numeratorIdx = numeratorIdx;
+      this._denominatorIdx = denominatorIdx;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        long levelValue = categoricalChunk.at8(i);
+        String factor = categoricalChunk.vec().factor(levelValue);
+
+        int foldValue = -1;
+        if(_foldColumnIdx != -1) {
+          long foldValueFromVec = cs[_foldColumnIdx].at8(i);
+          assert foldValueFromVec <= Integer.MAX_VALUE && foldValueFromVec >= 0 : FOLD_VALUE_IS_OUT_OF_RANGE_MSG;
+          foldValue = (int) foldValueFromVec;
+        }
+        _encodingDataMapPerNode.put(new CompositeLookupKey(factor, foldValue), new EncodingData(numeratorChunk.at8(i), denominatorChunk.at8(i)));
+      }
+    }
+
+    @Override
+    public void reduce(FrameWithEncodingDataToHashMap mrt) {
+      _encodingDataMapFinal.putAll(mrt.getEncodingDataMap());
```

Comment:
this will work but it is better to do 

_encodingDataMapPerNode.putAll(mrt.getEncodingDataMap())

because that minimizes necessary data transfer between nodes - _encodingDataMapPerNode is serialized as well as the MRTask result so data is partially duplicated in _encodingDataMapPerNode and _encodingDataMapFinal

Reply:
@michalkurka Could you please help me to understand it better.... Reduce phase will be performed on a single node once all map tasks are finished and sent back to a reduce node. It means that map nodes are sending only empty serialised hashmaps of `_encodingDataMapFinal` to the reducer. After reduce phase is completed we will end up with all data in `_encodingDataMapFinal` hashmap and some data in '_encodingDataMapPerNode" - and it is a partial duplicate of the data. But I don't understand where would we be sending unnesessary data between nodes during MapReduce. 
Did you mean those empty serialised maps are going to be send around or something else? Or maybe you meant that once reduce phase is finished and final map is filled it will be redistributed across all nodes? 
Actually I would like to  know if we wiat for all maps to finish before starting doing something (shuffling, reducing) ?

Reply:
@michalkurka I also found a bug in the test and probably in the code. Reduce method was not being called at all. Trying to understand why. I can't wrap my head around the comment of the MRTasl.reduce method. Is says `Reduce is optional if, e.g., the result is some output vector`.

Reply:
MRTask are always serialized as a whole, there is no difference between input and output - if you have 2 structures with duplicate data both structures will be serialized and send back - which is not necessary

in the hypothetical scenario of 2 nodes: the second node will do all the local reduces then serialized the MRTask send it to the first node and that will call the final reduce - because you have 2 structures populated in your task it will in worst case double the amount of data that need to be transferred

reduce is called always when there are actually data to reduce on, if the map is not applied reduce will not be called

Reply:
@michalkurka I think that in 1 node scenario reduce method is not supposed to be called and that is another reason not to use separate hashMap for final aggregation -  it would not be filled up in single node scenario bc reduce would not be called. With my distributed test... I think that setting size of cloud to 2 does not mean that task will be distributed, I need to force it to do it somehow. Maybe with layout or size of the frame.

Reply:
reduce will be called even in a single node scenario but you need at least 2 chunks - if there is nothing to reduce on (just a single map() invocation) it won't be called

you are right in your assessment that this is also a good reason not to have the second "final" structure

Reply:
I meant that it can have multiple reduces on 1 node but it is not splitting into multiple tasks in my test scenario.

Reply:
@michalkurka got it, thanks! I pushed commit with fixes. All single node tests are working but my `BroadcastJoinDistributedTest` is failing - I don't understand why. Maybe It is because of the improper layout ( but I tried different ones). I set it in a way that sum of them is equal to the number of rows. I can see in debug mode that reduce is being called twice as expected but in one call map is empty.

Reply:
@michalkurka have you had a chance to look into  ^^?  I was not able to see if it is passing or not on Jenkins as Junit tests support for AutoML that you have added is still not working(last time I checked January 30th). I was pushing Michal Raska month ago but this stuff was out of priority .... but I think we need to get back to this.